flink-runtime/src/main/java/org/apache/flink/runtime/client/JobClientActor.java

```diff
@@ -59,11 +87,50 @@ public JobClientActor(
 		this.leaderSessionID = Preconditions.checkNotNull(leaderSessionID, "The leader session ID option must not be null.");
 
 		this.sysoutUpdates = sysoutUpdates;
+		// set this to -1 to indicate the job hasn't been created yet.
+		this.currentJobCreatedAt = -1;
 	}
 	
 	@Override
 	protected void handleMessage(Object message) {
 		
+		// ======= Job status messages on regular intervals ==============
+		if(message instanceof JobManagerMessages.CurrentJobStatus){
+			JobStatus statusReport = ((JobManagerMessages.CurrentJobStatus) message).status();
+			long timeDiff;
+			switch(statusReport){
+				case RUNNING:
+					// Vincent, we happy?
+					this.currentJobCreatedAt = -1;
+					break;
+				case FINISHED:
+					// Yeah! We happy!
+					this.currentJobCreatedAt = -1;
+					break;
+				case CREATED:
+					// we're still at Job CREATED. Let's see if we're over the limit.
+					timeDiff = (System.currentTimeMillis() - this.currentJobCreatedAt);
+					if(timeDiff > JOB_CLIENT_JOB_STATUS_TIMEOUT){
+						failWithTimeout(timeDiff);
+					} // otherwise just wait a bit longer.
+					break;
+				case RESTARTING:
```

Comment:
What if a jobs gets restarted multiple times and we always miss the `RUNNING` state. Then it would look like as if the jobs stayed all the time in the `RESTARTING` state.